File: apps/ios/Sources/CrossyUI/RoomFactsSheet.swift
// The room-facts sheet (owner ruling 2026-07-10: the time pill is the room's
// facts; the tap-opened surface is a plain system sheet, 2026-07-12). A tap on
// the time pill mid-solve presents a small sheet, the ShareQRSheet register:
// the room's name, the live clock as the headline, the puzzle's facts, and (for
// the host) the end-game under a hairline, one confirm. The old inflate-from-the
// -pill morph (the PillInflation prototype: metaball / clean / overshoot) retired
// with this: the owner read it as ad-hoc goo, and a system sheet is the same
// verdict the players and share surfaces already earned (RosterMenu, ShareMenu).
//
// The sheet is a mid-solve surface only. At completion the pill seals and stands
// as the record (the frozen clock beside a check); the sheet does not auto-summon
// and a tap on a terminal pill does nothing (SolveScreen gates openFacts to
// ongoing). Post-game stats move to the clue-panel analysis surface, not here.
//
// The words and operations are pure (RoomFactsContent, FactsOperations); the
// clock ticks against `now` on the 1 Hz timeline, the bar clock's own arithmetic.

import CrossyDesign
import SwiftUI

/// The sheet's headline clock, one pure rule (pinned): the server's stat leads
/// when it exists (stats arrive only with `gameCompleted`, PROTOCOL.md §6);
/// otherwise the ambient clock's value, which ticks against `now` while the
/// room runs and freezes at the terminal instant (ID-2), exactly the bar
/// clock's own arithmetic.
public enum RoomFactsClock {
    public static func headline(
        solveTimeSeconds: Int?, firstFillAt: String?, completedAt: String?, now: Date
    ) -> String {
        if let solveTimeSeconds {
            return AmbientClock.display(seconds: solveTimeSeconds)
        }
        return AmbientClock.display(
            firstFillAt: firstFillAt, completedAt: completedAt, now: now)
    }
}

/// The sheet's words, derived once as plain strings so the view renders no
/// arithmetic (the StatsCardContent pattern). Mid-solve the label is the room's
/// name and the detail the puzzle's facts (title, byline, date). The completed
/// branch stays for the pure derivation's tests; the live sheet never opens on a
/// completed room (the pill seals instead), so in practice only the mid-solve
/// words render.
public struct RoomFactsContent: Equatable, Sendable {
    public let label: String
    public let detail: String?
    /// The room's check record while it runs ("Checked once" / "Checked N times",
    /// PROTOCOL.md §10, D27; design R10): a quiet, neutral line among the facts —
    /// no attribution, matching the wire event's missing `by`. Nil until the first
    /// accepted check. Mid-solve only: after completion the count freezes into
    /// `stats.checkCount`, whose display home is a future analysis-surface row.
    public let checkedLine: String?

    public init(label: String, detail: String?, checkedLine: String? = nil) {
        self.label = label
        self.detail = detail
        self.checkedLine = checkedLine
    }

    public static func make(
        roomName: String,
        puzzleTitle: String?,
        puzzleAuthor: String?,
        puzzleDate: String?,
        completed: Bool,
        totalEvents: Int?,
        participantCount: Int?,
        checkCount: Int = 0
    ) -> RoomFactsContent {
        if completed {
            var parts: [String] = []
            if let totalEvents {
                parts.append(totalEvents == 1 ? "1 entry" : "\(totalEvents) entries")
            }
            if let participantCount {
                parts.append(
                    participantCount == 1 ? "1 solver" : "\(participantCount) solvers")
            }
            return RoomFactsContent(
                label: RoomTerminal.completedNotice,
                detail: parts.isEmpty ? nil : parts.joined(separator: " · "))
        }
        let facts = [puzzleTitle, puzzleAuthor, puzzleDate]
            .compactMap { $0 }
            .filter { !$0.isEmpty }
        return RoomFactsContent(
            label: roomName,
            detail: facts.isEmpty ? nil : facts.joined(separator: " · "),
            checkedLine: Self.checkedLine(count: checkCount))
    }

    /// The R10 wording, natural casing, no zeros: nil before the first check.
    static func checkedLine(count: Int) -> String? {
        switch count {
        case ..<1: return nil
        case 1: return "Checked once"
        default: return "Checked \(count) times"
        }
    }
}

/// The sheet's operations, derived once so the view renders no policy (the
/// RoomFactsContent pattern). Two rows can stand: the room check (any host or
/// solver, PROTOCOL.md §5, §10; D27) above the host's end-game (`POST
/// /games/{id}/abandon`, host only, PROTOCOL.md §12). Both take a confirm step
/// in the view; a participant offered neither sees facts alone, which is fine.
public struct FactsOperations: Equatable, Sendable {
    /// The check row's render facts (design R7): present means the row stands;
    /// enabled only when the grid is full, with the quiet remaining-cells hint
    /// teaching the gate below full.
    public struct Check: Equatable, Sendable {
        /// Playable cells still empty in SEQUENCED state (R9: overlays excluded,
        /// the server's own filledCount gate mirrored).
        public let emptyCells: Int

        public init(emptyCells: Int) {
            self.emptyCells = max(0, emptyCells)
        }

        /// The grid-full gate (PROTOCOL.md §5: checkPuzzle requires a full grid).
        public var isEnabled: Bool { emptyCells == 0 }

        /// The quiet trailing hint while the grid is short; nil at full. Inside
        /// the row's standard height, no extra slot (R7).
        public var hint: String? {
            switch emptyCells {
            case 0: return nil
            case 1: return "1 empty"
            default: return "\(emptyCells) empty"
            }
        }
    }

    /// The check row, hosts and solvers only, and only where the transport
    /// carries `checkPuzzle` (R8: the demo's loopback drops it, so the demo
    /// never grows the row). Nil renders nothing.
    public let check: Check?
    /// The host's destructive end-game, offered only to the host (the server
    /// refuses a non-host abandon anyway; the client simply does not show it).
    public let canEndGame: Bool

    public init(check: Check? = nil, canEndGame: Bool) {
        self.check = check
        self.canEndGame = canEndGame
    }

    /// The operations for the local participant. `isHost` gates the destructive
    /// end-game; the check row needs a live check-capable transport (R8), a
    /// playing seat (spectators never see it, PROTOCOL.md §5's host|solver), and
    /// carries the sequenced empty-cell count for its own enable gate (R9).
    public static func make(
        isHost: Bool, isSpectator: Bool, supportsCheck: Bool, emptyCells: Int
    ) -> FactsOperations {
        FactsOperations(
            check: supportsCheck && !isSpectator ? Check(emptyCells: emptyCells) : nil,
            canEndGame: isHost)
    }

    /// The empty set (a spectator non-host, or any state that offers no operations).
    public static let none = FactsOperations(check: nil, canEndGame: false)

    /// How many operation rows render, the sheet-height arithmetic's input.
    public var rowCount: Int {
        (check != nil ? 1 : 0) + (canEndGame ? 1 : 0)
    }

    /// Whether the sheet renders a hairline and any operation rows at all.
    public var hasAny: Bool { rowCount > 0 }
}

/// The facts sheet's pure geometry, pinned in tests (the ShareQRSheetLayout
/// discipline): the detent height is slot arithmetic, never font metrics, so the
/// sheet sizes to exactly its content.
enum RoomFactsSheetLayout {
    static let verticalPadding: CGFloat = 32
    static let horizontalPadding: CGFloat = 24
    static let labelHeight: CGFloat = 18
    static let labelGap: CGFloat = 10
    /// The headline: the timer as the sheet's largest fact (ID-2).
    static let headlineFontSize: CGFloat = 52
    static let timeHeight: CGFloat = 60
    static let detailGap: CGFloat = 10
    static let detailHeight: CGFloat = 18
    /// The checked-count facts line (R10), its own slot (R7: the formula counts
    /// every conditional honestly or the sheet clips).
    static let checkedLineGap: CGFloat = 10
    static let checkedLineHeight: CGFloat = 18
    /// The operations block: air, a hairline, air, then the rows (check above
    /// end-game). The check row's remaining-cells hint renders INSIDE the standard
    /// row height at the trailing edge — no extra slot (R7).
    static let operationsAirAbove: CGFloat = 22
    static let dividerHeight: CGFloat = 1
    static let operationsAirBelow: CGFloat = 14
    static let operationRowHeight: CGFloat = 44

    static func height(
        hasDetail: Bool, hasCheckedLine: Bool, operationRows: Int
    ) -> CGFloat {
        verticalPadding * 2 + labelHeight + labelGap + timeHeight
            + (hasDetail ? detailGap + detailHeight : 0)
            + (hasCheckedLine ? checkedLineGap + checkedLineHeight : 0)
            + (operationRows > 0
                ? operationsAirAbove + dividerHeight + operationsAirBelow
                    + operationRowHeight * CGFloat(operationRows)
                : 0)
    }
}

/// The facts sheet: a short system sheet in the ShareQRSheet register (a fixed
/// detent, the drag indicator, the room's canvas to the safe area). Chrome stays
/// achromatic (§3): the sheet carries no color, only weight, and the one
/// destructive action shows its red in the system confirm dialog, never the body.
/// The 1 Hz timeline keeps the headline honest while the room runs.
@available(iOS 17.0, macOS 14.0, *)
@MainActor
struct RoomFactsSheet: View {
    let ground: GridGround
    let content: RoomFactsContent
    /// Already gated by the caller: the check row and the host's end-game
    /// mid-solve, `.none` otherwise.
    let operations: FactsOperations
    let solveTimeSeconds: Int?
    let firstFillAt: String?
    let completedAt: String?
    /// Check the puzzle for the room (PROTOCOL.md §5, §10; D27). Confirmed here
    /// first, then reported; the caller re-derives the grid-full gate at the
    /// confirm tap (design R2) and owns the send.
    let onCheckPuzzle: () -> Void
    /// End the game (host abandon). Confirmed here first, then reported.
    let onEndGame: () -> Void

    @State private var confirmingCheck = false
    @State private var confirmingEnd = false

    private var ink: Color { Color(rgb: ground.tokens.ink) }
    private var quiet: Color { Color(rgb: ground.tokens.number) }

    var body: some View {
        sheetBody
            #if os(iOS)
                .presentationDetents([
                    .height(
                        RoomFactsSheetLayout.height(
                            hasDetail: content.detail != nil,
                            hasCheckedLine: content.checkedLine != nil,
                            operationRows: operations.rowCount))
                ])
                .presentationDragIndicator(.visible)
            #endif
    }

    private var sheetBody: some View {
        TimelineView(.periodic(from: .now, by: 1)) { timeline in
            factsBody(now: timeline.date)
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(Color(rgb: ground.tokens.canvas).ignoresSafeArea())
        .confirmationDialog(
            "End this game for everyone?",
            isPresented: $confirmingEnd,
            titleVisibility: .visible
        ) {
            // The one destructive confirm (EXPERIENCE.md: abandon, one confirm,
            // plainly worded). The system dialog owns the red; the sheet body
            // stays achromatic (DESIGN.md §3).
            Button("End game", role: .destructive, action: onEndGame)
            Button("Keep playing", role: .cancel) {}
        } message: {
            Text(verbatim: "This ends the game for everyone in the room.")
        }
    }

    private func factsBody(now: Date) -> some View {
        let time = RoomFactsClock.headline(
            solveTimeSeconds: solveTimeSeconds,
            firstFillAt: firstFillAt, completedAt: completedAt, now: now)
        return VStack(alignment: .leading, spacing: 0) {
            facts(time: time)
            if operations.hasAny {
                operationBlock()
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(.vertical, RoomFactsSheetLayout.verticalPadding)
        .padding(.horizontal, RoomFactsSheetLayout.horizontalPadding)
    }

    // MARK: Facts (the label, the headline time, the quiet detail)

    private func facts(time: String) -> some View {
        VStack(alignment: .leading, spacing: 0) {
            // Natural casing: the uppercased small-caps register read wrong on
            // device (owner ruling 2026-07-10).
            Text(verbatim: content.label)
                .font(.system(size: 15, weight: .medium))
                .foregroundStyle(quiet)
                .lineLimit(1)
                .truncationMode(.tail)
                .frame(height: RoomFactsSheetLayout.labelHeight, alignment: .leading)
            Color.clear.frame(height: RoomFactsSheetLayout.labelGap)
            Text(verbatim: time)
                .font(.system(
                    size: RoomFactsSheetLayout.headlineFontSize, weight: .semibold))
                .monospacedDigit()
                .foregroundStyle(ink)
                .frame(height: RoomFactsSheetLayout.timeHeight, alignment: .leading)
            if let detail = content.detail {
                Color.clear.frame(height: RoomFactsSheetLayout.detailGap)
                Text(verbatim: detail)
                    .font(.system(size: 15, weight: .medium))
                    .monospacedDigit()
                    .foregroundStyle(quiet)
                    .lineLimit(1)
                    .truncationMode(.tail)
                    .frame(
                        height: RoomFactsSheetLayout.detailHeight, alignment: .leading)
            }
            if let checkedLine = content.checkedLine {
                // The check record (R10): quiet and neutral, no attribution
                // (the wire event carries no `by`, D27).
                Color.clear.frame(height: RoomFactsSheetLayout.checkedLineGap)
                Text(verbatim: checkedLine)
                    .font(.system(size: 15, weight: .medium))
                    .monospacedDigit()
                    .foregroundStyle(quiet)
                    .lineLimit(1)
                    .frame(
                        height: RoomFactsSheetLayout.checkedLineHeight,
                        alignment: .leading)
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(Text(verbatim: accessibilityLine(time: time)))
    }

    // MARK: Operations (only what the API already supports, §12)

    private func operationBlock() -> some View {
        VStack(alignment: .leading, spacing: 0) {
            Color.clear.frame(height: RoomFactsSheetLayout.operationsAirAbove)
            // A deterministic one-point hairline, not a system Divider: the
            // sheet's detent height is pinned arithmetic (RoomFactsSheetLayout),
            // and a hairline's scale-dependent height would unpin it.
            Rectangle()
                .fill(quiet.opacity(0.28))
                .frame(maxWidth: .infinity)
                .frame(height: RoomFactsSheetLayout.dividerHeight)
            Color.clear.frame(height: RoomFactsSheetLayout.operationsAirBelow)
            if let check = operations.check {
                checkRow(check)
            }
            if operations.canEndGame {
                operationRow(
                    "End game", systemImage: "xmark.circle",
                    action: { confirmingEnd = true })
            }
        }
    }

    /// The check row (PROTOCOL.md §5, §10; D27), above end-game: enabled only on
    /// a full grid, teaching the gate below full with the quiet remaining-cells
    /// hint at the trailing edge, inside the standard row height (R7). The
    /// confirm is the end-game register exactly but non-destructive (D27: the
    /// client interposes the one confirmation; the command is the confirmed
    /// intent) — plain tint, no red.
    private func checkRow(_ check: FactsOperations.Check) -> some View {
        Button(action: { confirmingCheck = true }) {
            HStack(spacing: 12) {
                Image(systemName: "checkmark.circle")
                    .font(.system(size: 16))
                    .frame(width: 24)
                Text(verbatim: "Check puzzle")
                    .font(.system(size: 16, weight: .medium))
                Spacer(minLength: 0)
                if let hint = check.hint {
                    Text(verbatim: hint)
                        .font(.system(size: 13, weight: .medium))
                        .monospacedDigit()
                        .foregroundStyle(quiet)
                }
            }
            .foregroundStyle(check.isEnabled ? ink : quiet)
            .frame(maxWidth: .infinity)
            .frame(height: RoomFactsSheetLayout.operationRowHeight)
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .disabled(!check.isEnabled)
        .confirmationDialog(
            "Check the puzzle for everyone?",
            isPresented: $confirmingCheck,
            titleVisibility: .visible
        ) {
            Button("Check puzzle", action: onCheckPuzzle)
            Button("Keep solving", role: .cancel) {}
        } message: {
            Text(verbatim: "Wrong letters get marked for the whole room.")
        }
    }

    private func operationRow(
        _ title: String, systemImage: String, action: @escaping () -> Void
    ) -> some View {
        // Achromatic like all chrome (DESIGN.md §3): the destructive row reads in
        // ink too, and the red lives in the system confirm dialog.
        Button(action: action) {
            HStack(spacing: 12) {
                Image(systemName: systemImage)
                    .font(.system(size: 16))
                    .frame(width: 24)
                Text(verbatim: title)
                    .font(.system(size: 16, weight: .medium))
                Spacer(minLength: 0)
            }
            .foregroundStyle(ink)
            .frame(maxWidth: .infinity)
            .frame(height: RoomFactsSheetLayout.operationRowHeight)
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
    }

    private func accessibilityLine(time: String) -> String {
        var line = "\(content.label), \(time)"
        if let detail = content.detail {
            line += ", \(detail.replacingOccurrences(of: " · ", with: ", "))"
        }
        if let checkedLine = content.checkedLine {
            line += ", \(checkedLine)"
        }
        return line
    }
}

File: apps/android/ui/src/main/kotlin/crossy/ui/RoomFactsSheet.kt
// The room-facts sheet (owner ruling 2026-07-10: the time pill is the room's facts; twin of apps/ios
// RoomFactsSheet.swift). A tap on the time pill mid-solve presents this sheet: the room's name, the
// live clock as the headline, the puzzle's facts when known, and (for the host) the End game action
// under a hairline, one two-beat confirm. A mid-solve surface only: at completion the pill seals and
// stands as the record, and the caller gates the summon to `ongoing`, so a tap on a terminal pill does
// nothing. The words and operations are pure (RoomFactsContent, FactsOperations, RoomFactsClock); the
// clock ticks against `now` on a 1 Hz loop, the bar clock's own arithmetic. A solid token surface (the
// glass register is a later material pass); the behavior is in.

package crossy.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.gestures.detectTapGestures
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlinx.coroutines.delay

/** The facts sheet's copy, greppable in one place (ID-5). The End game confirm is the one destructive
 *  two-beat (EXPERIENCE.md: abandon, one confirm, plainly worded). */
object RoomFactsCopy {
    const val endGameAction = "End game"
    const val endGameConfirmTitle = "End this game for everyone?"
    const val endGameConfirmBody = "This ends the game for everyone in the room."
    const val endGameConfirmAction = "End game"
    const val endGameCancelAction = "Keep playing"

    // The room check (PROTOCOL.md §5, §10; D27), word-for-word the iOS/web copy so the permanence
    // disclosure reaches every platform the same way. The confirm is the end-game register exactly
    // but non-destructive (a check is a room act, not a teardown): plain tint, no red.
    const val checkAction = "Check puzzle"
    const val checkConfirmTitle = "Check the puzzle for everyone?"
    const val checkConfirmBody = "Wrong letters get marked for the whole room. This is recorded."
    const val checkConfirmAction = "Check puzzle"
    const val checkCancelAction = "Keep solving"
}

/** The header's spoken line (twin of iOS RoomFactsSheet.accessibilityLine): the name, the clock, then
 *  the facts with the visual " · " separator read as a comma pause. */
internal fun factsAccessibilityLine(label: String, clock: String, detail: String?, checkedLine: String? = null): String {
    var line = "$label, $clock"
    if (detail != null) line += ", " + detail.replace(" · ", ", ")
    if (checkedLine != null) line += ", $checkedLine"
    return line
}

/** The sheet's headline clock, one pure rule (twin of iOS RoomFactsClock): the server's stat leads
 *  when it exists (stats arrive only with completion, PROTOCOL.md §6); otherwise the ambient clock's
 *  value, which ticks against `now` while the room runs and freezes at the terminal instant (ID-2). */
object RoomFactsClock {
    fun headline(
        solveTimeSeconds: Int?,
        firstFillAt: String?,
        freezeAt: String?,
        nowMillis: Long,
    ): String {
        if (solveTimeSeconds != null) return AmbientClock.display(solveTimeSeconds)
        return AmbientClock.display(firstFillAt, freezeAt, nowMillis)
    }
}

/** The sheet's words, derived once as plain strings (twin of iOS RoomFactsContent). Mid-solve the
 *  label is the room's name and the detail the puzzle's facts (title, byline, date), each dropped when
 *  absent. The wire types carry no puzzle metadata yet, so a live sheet's detail is usually null and
 *  the label alone stands. */
data class RoomFactsContent(val label: String, val detail: String?, val checkedLine: String? = null) {
    companion object {
        fun make(
            roomName: String,
            puzzleTitle: String? = null,
            puzzleAuthor: String? = null,
            puzzleDate: String? = null,
            // The room's mid-solve check record (PROTOCOL.md §10, D27; design R10): a quiet, neutral
            // line among the facts, no attribution (the wire event carries no `by`). Nil until the
            // first accepted check.
            checkCount: Int = 0,
            // The sitting count as facts context (owner ruling, D29): appended to the " · " grammar
            // only at two or more; one sitting (or a pre-D29 null) reads exactly as today.
            sittingCount: Int? = null,
        ): RoomFactsContent {
            val facts = buildList {
                addAll(listOfNotNull(puzzleTitle, puzzleAuthor, puzzleDate).filter { it.isNotEmpty() })
                if (sittingCount != null && sittingCount >= 2) add("$sittingCount sittings")
            }
            return RoomFactsContent(
                label = roomName,
                detail = if (facts.isEmpty()) null else facts.joinToString(" · "),
                checkedLine = checkedLine(checkCount),
            )
        }

        /** The R10 wording, natural casing, no zeros: null before the first check. */
        internal fun checkedLine(count: Int): String? = when {
            count < 1 -> null
            count == 1 -> "Checked once"
            else -> "Checked $count times"
        }
    }
}

/** The sheet's operations, derived once so the view renders no policy (twin of iOS FactsOperations).
 *  Two rows can stand: the room check (any host or solver, PROTOCOL.md §5, §10; D27) above the host's
 *  end-game (POST /games/{id}/abandon, host only, §12). A participant offered neither sees facts alone. */
data class FactsOperations(val check: Check?, val canEndGame: Boolean) {
    val hasAny: Boolean get() = check != null || canEndGame

    /** How many operation rows render (the check above end-game), the sheet-height input. */
    val rowCount: Int get() = (if (check != null) 1 else 0) + (if (canEndGame) 1 else 0)

    /** The check row's render facts (design R7): present means the row stands; enabled only when the
     *  grid is full, with the quiet remaining-cells hint teaching the gate below full. */
    data class Check(val emptyCells: Int) {
        // A negative input (a stand-in puzzle racing state) clamps to zero (R9: sequenced state only).
        val clampedEmpty: Int get() = maxOf(0, emptyCells)

        /** The grid-full gate (PROTOCOL.md §5: checkPuzzle requires a full grid). */
        val isEnabled: Boolean get() = clampedEmpty == 0

        /** The quiet trailing hint while the grid is short; null at full. Natural casing, singular at one. */
        val hint: String? get() = when (clampedEmpty) {
            0 -> null
            1 -> "1 empty"
            else -> "$clampedEmpty empty"
        }
    }

    companion object {
        /** `isHost` gates the destructive end-game; the check row needs a live check-capable transport
         *  (R8: the demo's loopback drops the command), a playing seat (spectators never see it,
         *  §5's host|solver), and carries the sequenced empty-cell count for its own enable gate (R9). */
        fun make(
            isHost: Boolean,
            isSpectator: Boolean,
            supportsCheck: Boolean,
            emptyCells: Int,
        ): FactsOperations = FactsOperations(
            check = if (supportsCheck && !isSpectator) Check(emptyCells) else null,
            canEndGame = isHost,
        )

        val none = FactsOperations(check = null, canEndGame = false)
    }
}

/**
 * The facts sheet as pixels. Presented by the composition root off the time pill's tap (gated to
 * ongoing). The 1 Hz loop keeps the headline honest while the room runs; [onEndGame] fires only after
 * the two-beat confirm.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun RoomFactsSheet(
    ground: GridGround,
    content: RoomFactsContent,
    operations: FactsOperations,
    solveTimeSeconds: Int?,
    firstFillAt: String?,
    freezeAt: String?,
    // Check the puzzle for the room (PROTOCOL.md §5, §10; D27). Confirmed here first, then reported;
    // the caller re-derives the grid-full gate at the confirm tap (design R2) and owns the send.
    onCheckPuzzle: () -> Unit = {},
    onEndGame: () -> Unit,
    onDismiss: () -> Unit,
) {
    val tokens = ground.tokens
    val sheetState = rememberModalBottomSheetState()
    var now by remember { mutableStateOf(System.currentTimeMillis()) }
    var confirmingCheck by remember { mutableStateOf(false) }
    var confirmingEnd by remember { mutableStateOf(false) }
    // The 1 Hz headline tick: the room's shared clock ages while the sheet stands (a completed
    // headline reads a fixed stat, so the tick is harmless there).
    LaunchedEffect(Unit) {
        while (true) {
            now = System.currentTimeMillis()
            delay(1000)
        }
    }

    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = sheetState,
        containerColor = tokens.canvas.toColor(),
        contentColor = tokens.ink.toColor(),
    ) {
        Column(
            modifier = Modifier.fillMaxWidth().padding(horizontal = 24.dp).padding(top = 8.dp, bottom = 32.dp),
        ) {
            val label = content.label.takeIf { it.isNotBlank() } ?: "Crossy"
            val clock = RoomFactsClock.headline(solveTimeSeconds, firstFillAt, freezeAt, now)
            // The facts read as one element (iOS accessibilityElement children:.ignore +
            // accessibilityLine "label, time, detail"); the End game row below stays its own node.
            Column(
                modifier = Modifier.fillMaxWidth().clearAndSetSemantics {
                    contentDescription = factsAccessibilityLine(label, clock, content.detail, content.checkedLine)
                },
            ) {
                Text(
                    label,
                    color = tokens.number.toColor(),
                    fontSize = 15.sp,
                    fontWeight = FontWeight.Medium,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                )
                Text(
                    clock,
                    color = tokens.ink.toColor(),
                    fontSize = 52.sp,
                    fontWeight = FontWeight.SemiBold,
                    style = androidx.compose.ui.text.TextStyle.Default.withTabularNumerals(),
                    modifier = Modifier.padding(top = 10.dp),
                )
                content.detail?.let { detail ->
                    Text(
                        detail,
                        color = tokens.number.toColor(),
                        fontSize = 15.sp,
                        fontWeight = FontWeight.Medium,
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis,
                        modifier = Modifier.padding(top = 10.dp),
                    )
                }
                // The check record (R10): quiet and neutral, no attribution (the wire event carries
                // no `by`, D27). Its own line below the facts detail, tabular so the count sits still.
                content.checkedLine?.let { checkedLine ->
                    Text(
                        checkedLine,
                        color = tokens.number.toColor(),
                        fontSize = 15.sp,
                        fontWeight = FontWeight.Medium,
                        maxLines = 1,
                        style = androidx.compose.ui.text.TextStyle.Default.withTabularNumerals(),
                        modifier = Modifier.padding(top = 10.dp),
                    )
                }
            }
            // The operations block (the check above end-game, R7): air, one hairline, air, then the
            // rows, rendered only when at least one stands. The check row's remaining-cells hint sits
            // INSIDE the standard row height at the trailing edge, no extra slot.
            if (operations.hasAny) {
                Box(
                    Modifier
                        .fillMaxWidth()
                        .padding(top = 22.dp, bottom = 14.dp)
                        .height(1.dp)
                        .background(tokens.number.toColor().copy(alpha = 0.28f)),
                )
                operations.check?.let { check ->
                    Row(
                        modifier = Modifier
                            .fillMaxWidth()
                            .height(44.dp)
                            .then(
                                if (check.isEnabled) {
                                    Modifier.pointerInput(Unit) { detectTapGestures { confirmingCheck = true } }
                                } else {
                                    Modifier
                                },
                            ),
                        verticalAlignment = Alignment.CenterVertically,
                    ) {
                        Text(
                            RoomFactsCopy.checkAction,
                            // The gate teaches through tone: full is ink, short is the quiet number
                            // color (disabled), never red (a check is non-destructive, D27).
                            color = if (check.isEnabled) tokens.ink.toColor() else tokens.number.toColor(),
                            fontSize = 16.sp,
                            fontWeight = FontWeight.Medium,
                        )
                        check.hint?.let { hint ->
                            Spacer(Modifier.weight(1f))
                            Text(
                                hint,
                                color = tokens.number.toColor(),
                                fontSize = 13.sp,
                                fontWeight = FontWeight.Medium,
                                style = androidx.compose.ui.text.TextStyle.Default.withTabularNumerals(),
                            )
                        }
                    }
                }
                if (operations.canEndGame) {
                    Row(
                        modifier = Modifier
                            .fillMaxWidth()
                            .height(44.dp)
                            .pointerInput(Unit) { detectTapGestures { confirmingEnd = true } },
                        verticalAlignment = Alignment.CenterVertically,
                    ) {
                        Text(
                            RoomFactsCopy.endGameAction,
                            color = tokens.ink.toColor(),
                            fontSize = 16.sp,
                            fontWeight = FontWeight.Medium,
                        )
                    }
                }
            }
        }
    }

    if (confirmingCheck) {
        AlertDialog(
            onDismissRequest = { confirmingCheck = false },
            containerColor = tokens.canvas.toColor(),
            title = { Text(RoomFactsCopy.checkConfirmTitle, color = tokens.ink.toColor()) },
            text = { Text(RoomFactsCopy.checkConfirmBody, color = tokens.number.toColor()) },
            confirmButton = {
                TextButton(onClick = {
                    confirmingCheck = false
                    onCheckPuzzle()
                    onDismiss()
                }) { Text(RoomFactsCopy.checkConfirmAction) }
            },
            dismissButton = {
                TextButton(onClick = { confirmingCheck = false }) { Text(RoomFactsCopy.checkCancelAction) }
            },
        )
    }

    if (confirmingEnd) {
        AlertDialog(
            onDismissRequest = { confirmingEnd = false },
            containerColor = tokens.canvas.toColor(),
            title = { Text(RoomFactsCopy.endGameConfirmTitle, color = tokens.ink.toColor()) },
            text = { Text(RoomFactsCopy.endGameConfirmBody, color = tokens.number.toColor()) },
            confirmButton = {
                TextButton(onClick = {
                    confirmingEnd = false
                    onEndGame()
                    onDismiss()
                }) { Text(RoomFactsCopy.endGameConfirmAction) }
            },
            dismissButton = {
                TextButton(onClick = { confirmingEnd = false }) { Text(RoomFactsCopy.endGameCancelAction) }
            },
        )
    }
}
